qubes/vm/dispvm.py

```diff
@@ -187,11 +189,130 @@ def __init__(self, app, xml, *args, **kwargs):
+
+        Template from which the VM should be created.
+
+        :return:
+        """
+        preload_dispvm_max = int(
+            self.features.check_with_template("preload-dispvm-max", 0)
+        )
+        if preload_dispvm_max == 0:
+            return
+
+        preload_dispvm = self.features.check_with_template(
+            "preload-dispvm", None
+        )
+        if preload_dispvm:
+            if len(preload_dispvm.split(" ")) >= preload_dispvm_max:
+                return
+            preload_dispvm += " " + self.name
+        else:
+            preload_dispvm = self.name
+
+        dispvm_template = getattr(self, "template")
+        dispvm_template.features["preload-dispvm"] = preload_dispvm
+        self.features["internal"] = True
+
+    async def use_preloaded_dispvm(self):
+        """
+        Mark preloaded DispVM as used.
+
+        Template from which the VM should be used (first preloaded VM) or the
+        VM itself.
+
+        :return:
+        """
+        if self.klass == "DispVM":
+            dispvm_template = getattr(self, "template")
+            dispvm = self
+        elif (
+            self.klass == "AppVM"
+            and getattr(self, "template_for_dispvms", False)
+        ):
+            dispvm_template = dispvm
+        else:
+            raise qubes.exc.QubesException("Invalid argument provided")
+
+        preload_dispvm = dispvm_template.features.get("preload-dispvm", None)
+        if not preload_dispvm:
+            return
+
+        preload_dispvm = preload_dispvm.split(" ")
+        if self.klass == "DispVM":
+            if dispvm.name not in preload_dispvm:
+                raise qubes.exc.QubesException("DispVM is not preloaded")
+            used_dispvm = dispvm.name
+        else:
+            used_dispvm = preload_dispvm[0]
+            dispvm = self.app.domains[used_dispvm]
+
+        unused_dispvm = preload_dispvm.remove(used_dispvm)
+        dispvm_template.features["preload-dispvm"] = unused_dispvm
+        dispvm.features["internal"] = False
+        ## TODO: confirm if this guess of event code is correct
+        self.app.fire_event_for_permission(dispvm_template=dispvm_template)
+        dispvm_template.fire_event_async("domain-preloaded-dispvm-used")
+
+    @qubes.events.handler(
+        "domain-preloaded-dispvm-used", "domain-preloaded-dispvm-autostart"
+    )
+    async def on_domain_preloaded_dispvm_used(self, event, delay=5):
+        """When preloaded DispVM is used or after boot, preload another one."""  # pylint: disable=unused-argument
+        ## TODO: fire event after start of all qubes that set to autostart.
+        await asyncio.sleep(delay)
+        while True:
+            ## TODO: Is there existing Qubes code that checks available memory
+            ##       before starting a qube?
+            memory = getattr(self, "memory", 0)
+            available_memory = (
+                psutil.virtual_memory().available / (1024 * 1024)
```

Comment:
I get what you mean, but this will not work. This looks only at free memory in dom0, not the whole system. And even if it would look more globally, qmemman tries to allocate available memory as much as possible. Only qmemman knows how much "free" memory you really have, and currently there is no API to query that...